Bug 1796658: Fix catalog tile truncation and responsiveness

rebeccaalpert opened the pull request:
Upgraded to latest version of catalog-view-extensions. Updated catalog tile sizing and grids to work with update, including the skeleton catalog. This should fix the issues @spadgett and @serenamarie125 noted with  prior work, notably poor truncation and responsive catalog tiles. Catalog tiles are now fixed width and much shorter. Text is truncated at one line (for installed OperatorHub tiles) and three maximum for all other tiles.

<img width="1383" alt="Screen Shot 2020-02-03 at 6 24 10 PM" src="https://user-images.githubusercontent.com/7014965/73699693-789e8880-46b2-11ea-8c84-332ef4a4d807.png">

<img width="1395" alt="Screen Shot 2020-02-03 at 6 24 36 PM" src="https://user-images.githubusercontent.com/7014965/73699705-82c08700-46b2-11ea-94b8-d35aabc358c2.png">

<img width="1379" alt="Screen Shot 2020-02-03 at 6 27 25 PM" src="https://user-images.githubusercontent.com/7014965/73699830-dcc14c80-46b2-11ea-882a-7f47f6265fe4.png">

OperatorHub Installed vs. Uninstalled:
<img width="797" alt="Screen Shot 2020-02-03 at 6 28 37 PM" src="https://user-images.githubusercontent.com/7014965/73699908-12fecc00-46b3-11ea-95c2-b3731b75a13e.png">

Example of new heights across rows with differing levels of truncation:
<img width="790" alt="Screen Shot 2020-02-03 at 6 28 50 PM" src="https://user-images.githubusercontent.com/7014965/73699918-18f4ad00-46b3-11ea-80b8-ad80c3a201f2.png">

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1796658 a.k.a. https://issues.redhat.com/browse/CONSOLE-2013.

Heads up @openshift/team-ux-review and @jcaianirh.

rebeccaalpert commented:
/retest

spadgett commented:
/assign @rhamilto

rhamilto commented on frontend/packages/operator-lifecycle-manager/src/style.scss:
```diff
@@ -12,7 +12,8 @@
 
 .co-catalog-tile {
-  height: 290px;
+  height: 100%;
```
The cards now have different heights.
height is 203px
![Screen Shot 2020-02-04 at 2 03 49 PM](https://user-images.githubusercontent.com/895728/73777434-69bde180-4757-11ea-8047-80941f6917bb.png)
height is 224px
![Screen Shot 2020-02-04 at 2 03 59 PM](https://user-images.githubusercontent.com/895728/73777435-69bde180-4757-11ea-9f87-5d4443534e38.png)

rebeccaalpert replied:
This was a recommendation from one of the PatternFly CSS devs. If we have fixed heights on these cards, we run into either ugly overflow bottom padding clipping or huge cards because we have to account for the largest possible card. The title and provider varies by as much as 1-5 lines that I've seen. With this method, the heights are consistent across the row and we avoid both of the fixed height issues.

rhamilto commented on frontend/public/components/utils/_skeleton-screen.scss:
```diff
@@ -42,7 +42,7 @@ $skeleton-overview-title-position: 50px $skeleton-overview-tile-padding;
 $skeleton-overview-title-width: 38%;
 
-$skeleton-tile-height: 290px; // height of catalog tiles
+$skeleton-tile-height: 240px; // height of catalog tiles
```
The height of the skeleton tile does not match the height of the cards because the cards can have different heights (see earlier comment).

rebeccaalpert replied:
Yeah, this was an issue. We obviously don't have data yet at this point to make them match the real tiles, which are based on the size of the data. I sized them to match the top few so it wouldn't be jarring, but if you have thoughts on a better approach, let us know.

benjaminapetersen replied:
Can use some `var + var` incantation instead of hard-coded `px`?

rhamilto commented on frontend/public/components/utils/_skeleton-screen.scss:
```diff
@@ -94,7 +94,7 @@ $skeleton-tile-desc-line-bone: linear-gradient(#fff $skeleton-tile-desc-line-hei
   display: flex;
   flex-wrap: wrap;
-  margin: 117px 6px 6px 0;
+  margin: 100px ($grid-gutter-width / 2) 0 0;
```
The skeleton cards now render higher up the page than the cards.

![JgCmmYxHxa](https://user-images.githubusercontent.com/895728/73778428-33816180-4759-11ea-984c-f84664c31722.gif)

rebeccaalpert replied:
This is a good catch; I'll fix this.

rebeccaalpert commented on frontend/yarn.lock:
```diff
@@ -1233,7 +1233,7 @@
     xterm "^3.3.0"
 
-"@patternfly/react-core@3.124.1", "@patternfly/react-core@^3.124.1":
+"@patternfly/react-core@3.124.1", "@patternfly/react-core@^3.124.1", "@patternfly/react-core@^3.136.2":
```
Not sure why this got introduced; I'll take a look.

rebeccaalpert commented:
Fixed alignment issue and refactored to use global variables.

rebeccaalpert commented:
/retest

rhamilto commented:
/hold

please squash

rhamilto commented:
/lgtm cancel

rebeccaalpert commented:
Squashed.

spadgett commented:
/retest

rhamilto commented:
/retest

rhamilto commented:
Hmmm.  It looks like the PatternFly bumps are significantly increasing the size of vendor main bundle by 2.3 megs.  @spadgett?

rhamilto commented:
Looks like react-icons is the main culprit.

spadgett commented:
I think the analyze test flaked. We've seen that before unfortunately and need to investigate :/

spadgett commented:
/hold cancel

rhamilto commented:
/hold

spadgett commented:
I'm seeing some of the tiles getting cut off:

![Screen Shot 2020-02-18 at 3 36 44 PM](https://user-images.githubusercontent.com/1167259/74775881-8df3e500-5264-11ea-911a-88c36c9424e1.png)

Also it looks like bumping patternfly has caused layout issues for the notification drawer:

![Screen Shot 2020-02-18 at 3 36 23 PM](https://user-images.githubusercontent.com/1167259/74775931-a8c65980-5264-11ea-8ef5-c1e7d2f250a3.png)

cc @jcaianirh

rhamilto commented:
/lgtm cancel

jcaianirh commented:
> Also it looks like bumping patternfly has caused layout issues for the notification drawer:
> 
> cc @jcaianirh
Thanks @spadgett and @rebeccaalpert , this has been investigated, and a known patternfly-core issue introduced for master/detail has caused a problem with the drawer component in react. This breaking change has been backed out, and the react component will be updated this sprint to correctly handle the new html/css modifications. Upgrading the this pr to the next pf release will fix the break in behavior you see above.

spadgett commented:
/bugzilla refresh

rebeccaalpert commented:
I upgraded PF React, but I'm still seeing the expand-on-close issue on smaller viewports as well as the endless tile issue on very large viewports (sorry for missing this; I don't have a large monitor at my primary workspace). It looks like the default behavior of the drawer changes at smaller viewports and is open by default, but the pf-m-expanded class is not applied. For some reason, it is applied only on close. I reached out some PatternFly folks to investigate this.

The overflow issue appears to be caused by a margin that's applied only when the drawer is expanded:
```
.pf-c-drawer.pf-m-expanded.pf-m-inline {
*      --pf-c-drawer__content--MarginRight: var(--pf-c-drawer--m-expanded--m-inline__content--MarginRight);
```

Removing the .pf-m-expanded selector restores the expected behavior. I'll reach out to PatternFly on this as well.

spadgett commented:
/bugzilla refresh

rebeccaalpert commented:
My CSS changes didn't seem to be affecting the drawer or layout. I upgraded @patternfly/patternfly and the issues seemed to resolve (I also updated React, but didn't see much of a change there). I added a z-index style to the drawer to fix an issue with charts and filter bars I noticed. Let me know if you see anything else that seems odd.

rebeccaalpert commented:
```
could not wait for build: the build test-bin failed after 6m49s with reason DockerBuildFailed: Docker build strategy has failed.  [16] (webpack)/buildin/global.js 472 bytes {0} [built]         + 31 hidden modules error Command failed with exit code 2. info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command. error: build error: running '/bin/bash -c set -o errexit; umask 0002; ./build.sh' failed with exit code 2
```

/retest

rebeccaalpert commented:
```
could not wait for build: the build test-bin failed after 7m28s with reason DockerBuildFailed: Docker build strategy has failed.

[16] (webpack)/buildin/global.js 472 bytes {0} [built]
        + 31 hidden modules
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error: build error: running '/bin/bash -c set -o errexit; umask 0002; ./build.sh' failed with exit code 2
```

/retest

rebeccaalpert commented:
/retest

rebeccaalpert commented:
/retest

rebeccaalpert commented:
/retest

rebeccaalpert commented:
/retest

rebeccaalpert commented:
/retest

rebeccaalpert commented:
```
creates a new resource instance
   Error: Failed: omg js error: true
       at /go/src/github.com/openshift/console/frontend/integration-tests/protractor.conf.ts:234:7
       at ManagedPromise.invokeCallback_ /go/src/github.com/openshift/console/frontend/node_modules/selenium-webdriver/lib/promise.js:1376:14
       at TaskQueue.execute_ /go/src/github.com/openshift/console/frontend/node_modules/selenium-webdriver/lib/promise.js:3084:14
       at TaskQueue.executeNext_ /go/src/github.com/openshift/console/frontend/node_modules/selenium-webdriver/lib/promise.js:3067:27
```

/retest

rebeccaalpert commented:
/retest

rhamilto commented:
@rebeccaalpert, as always, please squash.

rebeccaalpert commented:
/retest

rebeccaalpert commented:
/retest

jcaianirh commented:
/test e2e-gcp-console

jcaianirh commented:
/test analyze

jcaianirh commented:
/test e2e-gcp-console

spadgett commented:
@jcaianirh fyi, this looks like it might be a legitimate test failure

jcaianirh commented:
@spadgett, tested this locally...getting the same error. Its a failure when creating a pod, though that code isn't associated with the changes in the pr. Also, the screenshot log output from the test looks accurate.

spadgett commented:
If we can reproduce locally, you might add some logging here to see what the error is:

https://github.com/rebeccaalpert/console/blob/20b5545f1b535f91088f78b959fa39e0b3f7e5d5/frontend/public/components/app.jsx#L178

spadgett commented:
```
Uncaught error TypeError: document.querySelector(...) is null backend.js:6:2315
    r backend.js:6
    onunhandledrejection app.jsx:180
TypeError: document.querySelector(...) is nulledit-yaml.jsx:223
    focusEditor edit-yaml.jsx:223
```

rebeccaalpert commented:
Thank you for figuring this out. I guess it's not finding the monaco-editor textarea? I'll take a look.

benjaminapetersen commented on frontend/public/components/app.jsx:
```diff
@@ -175,7 +175,11 @@ fetchSwagger();
 window.windowError = false;
-window.onerror = window.onunhandledrejection = () => (window.windowError = true);
+window.onerror = window.onunhandledrejection = (e) => {
+  // eslint-disable-next-line no-console
```
+1

spadgett commented:
> Thank you for figuring this out. I guess it's not finding the monaco-editor textarea? I'll take a look.

I suspect the PF popover is calling `onHide` when it didn't use to. Patternfly bug?

jcaianirh commented:
@spadgett thanks so much for finding that!

spadgett commented:
/test analyze

jcaianirh commented:
/test analyze

rebeccaalpert commented:
/cherry-pick release-4.4

rhamilto commented:
It appears the PatternFly bugs have been fixed.  \o/

I did noticed one new minor cosmetic issue:  the gray background overflows the "frame" at mobile in all browsers.

![localhost_9000_operatorhub_ns_robb(iPhone 5_SE)](https://user-images.githubusercontent.com/895728/75784596-715daf80-5d30-11ea-963c-b5623ced3727.png)

I'll let @benjaminapetersen decide if we want this to merge with the new bug.

rebeccaalpert commented:
Opened Bugzilla to track: https://bugzilla.redhat.com/show_bug.cgi?id=1809812

@benjaminapetersen said he's ok to sign off as long as we're tracking that issue.

rhamilto commented:
/lgtm

rebeccaalpert commented:
Got ok to cancel hold per @rhamilto.

/hold cancel

benjaminapetersen commented:
👍 
/lgtm

benjaminapetersen commented:
/cherry-pick release-4.4
